workers/unity/Packages/io.improbable.gdk.mobile/Editor/AndroidLaunchUtils.cs

```diff
@@ -1,29 +1,92 @@
+using System.Collections.Generic;
 using System.IO;
+using System.Text.RegularExpressions;
 using Improbable.Gdk.Tools;
 using UnityEditor;
 using UnityEngine;
 using Debug = UnityEngine.Debug;
 
 namespace Improbable.Gdk.Mobile
 {
-    // TODO UTY-2095: Improve Android workflow and refactor the code
     public static class AndroidLaunchUtils
     {
-        public static void Launch(bool shouldConnectLocally, string runtimeIp)
+        private static readonly Regex DeviceMatchRegex = new Regex("(?:(?<id>[\\w_\\d-]+)\\s*device).*"
+            + "(?:product:(?<product>[\\w_\\d]+))\\s*"
+            + "(?:model:(?<model>[\\w_\\d]+))\\s*"
+            + "(?:device:(?<device>[\\w_\\d]+)).*");
+
+        public static (Dictionary<string, string>, Dictionary<string, string>) RetrieveAvailableEmulatorsAndDevices()
+        {
+            var available = false;
+            var availableEmulators = new Dictionary<string, string>();
+            var availableDevices = new Dictionary<string, string>();
+
+            if (!TryGetAdbPath(out var adbPath))
+            {
+                Debug.LogError(
+                    $"Could not find Android SDK. Please set the SDK location in your editor preferences.");
+                return (availableEmulators, availableDevices);
+            }
+
+            // List connected devices
+            // adb devices -l
+            var result = RedirectedProcess.Command(adbPath)
+                .WithArgs("devices", "-l")
+                .AddOutputProcessing(message =>
+                {
+                    if (!DeviceMatchRegex.IsMatch(message))
+                    {
+                        return;
+                    }
+
+                    available = true;
+
+                    var match = DeviceMatchRegex.Match(message);
+                    var deviceId = match.Groups["id"].Value;
+
+                    if (deviceId.Contains("emulator"))
+                    {
+                        availableEmulators[$"{match.Groups["product"].Value} ({deviceId})"] = deviceId;
+                    }
+                    else
+                    {
+                        availableDevices[$"{match.Groups["model"].Value} ({deviceId})"] = deviceId;
+                    }
+                })
+                .RedirectOutputOptions(OutputRedirectBehaviour.None)
+                .Run();
+
+            if (result.ExitCode == 0)
+            {
+                if (!available)
+                {
+                    Debug.Log("No Android emulators or devices found.");
+                }
+
+                return (availableEmulators, availableDevices);
+            }
+
+            Debug.LogError("Failed to find Android emulators or devices.");
+            availableEmulators.Clear();
+            availableDevices.Clear();
+            return (availableEmulators, availableDevices);
+        }
+
+        public static void Launch(bool shouldConnectLocally, string deviceId, string runtimeIp, bool useEmulator)
```

Comment:
Not a huge fan of the bools here. Could do a little refactoring and make some intermediate data structures:

```csharp
internal struct MobileLaunchOptions 
{
	public bool ShouldConnectLocally;
	public string RuntimeIp;
	
	// If you move this in here, you can drop the `MobileLaunchUtils` class entirely.
	public string ToArgs()
	{
		...
	}
}

internal struct MobileDeviceTarget
{
	public string DeviceId;
	public bool IsEmulator;
	
	// Maybe rename this..? 
	public string DeviceOrEmulator => IsEmulator ? "emulator" : "device";
}

public static void Launch(MobileDeviceTarget target, MobileLaunchOptions options)
{
	...
}
```